packages/bug-detectors/internal/prototype-pollution.ts

```diff
@@ -0,0 +1,565 @@
+					);
+
+					const variable = path.node.id as types.Identifier;
+					instrumentationGuard.add(
+						TrackedExpressions.VARIABLE_DECLARATION,
+						markedDeclarator
+					);
+					if (types.isAssignmentExpression(path.node.init))
+						instrumentationGuard.add(
+							TrackedExpressions.ASSIGNMENT_EXPRESSION,
+							path.node.init
+						);
+
+					path.node.init = types.callExpression(
+						types.arrowFunctionExpression(
+							[],
+							types.blockStatement([
+								types.variableDeclaration("const", [markedDeclarator]),
+								// check for __proto__ changes
+								types.expressionStatement(
+									types.callExpression(
+										types.identifier(
+											"PrototypePollution.detectPrototypePollution"
+										),
+										[newVariable, types.stringLiteral("" + variable.name)]
+									)
+								),
+								// return the original initializer
+								types.returnStatement(newVariable),
+							])
+						),
+						[]
+					);
+					instrumentationGuard.add(
+						TrackedExpressions.VARIABLE_DECLARATION,
+						path.node.init
+					);
+				}
+			},
+		},
+	};
+});
+
+// These objects will be used to detect prototype pollution.
+// Using global arrays for performance reasons.
+const BASIC_OBJECTS = [
+	{},
+	[],
+	"",
+	42,
+	true,
+	() => {
+		/**/
+	},
+];
+// The names are used in the Findings to print nicer messages.
+const BASIC_OBJECT_NAMES = [
+	"Object",
+	"Array",
+	"String",
+	"Number",
+	"Boolean",
+	"Function",
+];
+
+// Compute prototype snapshots of each selected basic object before any fuzz tests are run.
+// These snapshots are used to detect prototype pollution after each fuzz test.
+const BASIC_PROTO_SNAPSHOTS = computeBasicPrototypeSnapshots();
+
+function detectPrototypePollutionOfBasicObjects(): void {
+	const currentProtoSnapshots = computeBasicPrototypeSnapshots();
+	// Compare the current prototype snapshots of basic objects to the original ones.
+	for (let i = 0; i < BASIC_PROTO_SNAPSHOTS.length; i++) {
+		if (!currentProtoSnapshots[i]) {
+			reportFinding(
+				`Prototype Pollution: prototype of ${BASIC_OBJECT_NAMES[i]} changed.`
+			);
+			continue;
+		}
+		const equalityResult = protoSnapshotsEqual(
+			BASIC_PROTO_SNAPSHOTS[i],
+			currentProtoSnapshots[i]
+		);
+		if (!equalityResult.equal) {
+			reportFinding(
+				`Prototype Pollution: prototype of ${BASIC_OBJECT_NAMES[i]} changed.${equalityResult.message}`
+			);
+		}
+	}
+}
+
+registerAfterEachCallback(detectPrototypePollutionOfBasicObjects);
+
+// There are two main ways to pollute a prototype of an object:
+// 1. Changing a prototype's property using __proto__
+// 2. Changing a prototype's property using constructor.prototype
+// This dictionary adds these strings to the fuzzer dictionary.
+// Adding strings targeted at specific protocols (XML, HTTP, protobuf, etc.) will reduce the performance of the fuzzer,
+// because it will try strings from the wrong protocol. Therefore, it is advised to add protocol-specific strings
```

Comment:
I don't quite understand the last paragraph: If protocol-specific strings are added to the dictionary for each fuzz test, doesn't that still risk reducing the fuzzer performance? 

In this particular case we could argue that these strings aren't really protocol-specific though as they come from the language itself.